pya/Aserver.py

```diff
@@ -73,69 +69,63 @@ def __init__(self, sr=44100, bs=256, device=None, channels=2, format=pyaudio.paF
-        self.pa = pyaudio.PyAudio()
+        if backend is None:
+            from .backend.PyAudio import PyAudioBackend
+            self.backend = PyAudioBackend(**kwargs)
+        else:
+            self.backend = backend
         self.channels = channels
-        self._status = pyaudio.paComplete
         self._device = 1
 
         # Get audio devices to input_device and output_device
         self.input_devices = []
         self.output_devices = []
-        for i in range(self.pa.get_device_count()):
-            if self.pa.get_device_info_by_index(i)['maxInputChannels'] > 0:
-                self.input_devices.append(self.pa.get_device_info_by_index(i))
-            if self.pa.get_device_info_by_index(i)['maxOutputChannels'] > 0:
-                self.output_devices.append(self.pa.get_device_info_by_index(i))
+        for i in range(self.backend.get_device_count()):
+            if self.backend.get_device_info_by_index(i)['maxInputChannels'] > 0:
+                self.input_devices.append(self.backend.get_device_info_by_index(i))
+            if self.backend.get_device_info_by_index(i)['maxOutputChannels'] > 0:
+                self.output_devices.append(self.backend.get_device_info_by_index(i))
 
         if device is None:
-            self.device = self.pa.get_default_output_device_info()['index']
+            self.device = self.backend.get_default_output_device_info()['index']
         else:
             self.device = device
 
-        # self.device_dict = self.pa.get_device_info_by_index(self.device)
+        # self.device_dict = self.backend.get_device_info_by_index(self.device)
         # # self.max_out_chn is not that useful: there can be multiple devices having the same mu
         # self.max_out_chn = self.device_dict['maxOutputChannels']
         # self.max_in_chn = self.device_dict['maxInputChannels']
 
-        self.format = format
         self.gain = 1.0
         self.srv_onsets = []
         self.srv_asigs = []
         self.srv_curpos = []  # start of next frame to deliver
         self.srv_outs = []  # output channel offset for that asig
-        self.pastream = None
-        self.dtype = 'float32'  # for pyaudio.paFloat32
-        self.range = 1.0
+        self.stream = None
         self.boot_time = None  # time.time() when stream starts
         self.block_cnt = None  # nr. of callback invocations
         self.block_duration = self.bs / self.sr  # nominal time increment per callback
         self.block_time = None  # estimated time stamp for current block
         self._stop = True
-        if self.format == pyaudio.paInt16:
-            self.dtype = 'int16'
-            self.range = 32767
-        if self.format not in [pyaudio.paInt16, pyaudio.paFloat32]:
-            warn(f"Aserver: currently unsupported pyaudio format {self.format}")
-        self.empty_buffer = np.zeros((self.bs, self.channels), dtype=self.dtype)
+        self.empty_buffer = np.zeros((self.bs, self.channels), dtype=self.backend.dtype)
```

Comment:
Q: if backend is not none, it supposedly would be a string such as 'file', which is above assigned to backend, so self.backend.dtype would not exist - unless each legal type would be checked above and self.backend set to a backend stream... - is this the intended solution?

Reply:
> it supposedly would be a string
that's one option but not necessarily the best one. As I mentioned before I am not a big fan of strings as a parameter to trigger other actions. In times of omnipresent code completion an enum might be more suitable. Currently the parameter is expected to be a backend or `None` (e.g. `Aserver(..., backend=DummyBackend(), ...)`). This keeps the coupling of `Aserver` and potential backends minimal. This means that `**kwargs` is only used in the `None` case but I don't think passing an instance is less convenient than passing the parameters AND a string/enum to `Aserver`:

```
# let's assume format and flag are backend parameters
Aserver(format=..., flag=...)  # 1) will be passed to PyAudioBackend in Aserver
Aserver(backend='pyaudio', format=..., flag=...)  # 2) resolve string and create backend in Aserver
Aserver(backend=PyAudioBackend(format=..., flags=...))  # 3) just pass the backend instance
```

1) has to be supported because of history reasons. I don't think 2) is much more convenient than 3).